contributors/devel/sig-instrumentation/migration-to-structured-logging.md

```diff
@@ -0,0 +1,329 @@
+Names should be picked based on semantic meaning of value itself, not the context in which is used (log message should
+imply the context). For example names like `status` should be used over (`desiredStatus`, `oldStatus`, `badStatus`) thus
+allowing to query and join different logs lines on `status` field.
+
+Kubernetes objects should be referenced using only their kind, no matter their api group or version. Example argument
+names: `deployment`, `pod`, `node`, `replicaSet`.
+
+In situations when we want to log value of same meaning twice (e.g. transition between state) it is ok to name
+additional argument based on context, but leaving one main value with name unchanged.
+
+Example:
+
+```go
+func LogStatus(oldStatus, desiredStatus, newStatus string) {
+   klog.InfoS("Status changed", "status", newStatus, "oldStatus", oldStatus, "desiredStatus", desiredStatus)
+}
+```
+
+[lowerCamelCase]: https://en.wiktionary.org/wiki/lowerCamelCase
+
+### Use `klog.KObj` and `klog.KRef` for Kubernetes objects
+
+As part of structured logging migration we want to ensure that kubernetes objects references are consistent within
+codebase. Two new utility functions were introduced to klog `klog.KObj` and `klog.KRef`. Any reference
+(name, uid, namespace) to Kubernetes Object (Pod, Node, Deployment, CRD) should be rewritten to utilize those functions.
+
+For example
+```go
+func updatePod(pod *covev1.Pod) {
+   ...
+   klog.Infof("updated pod %s in namespace %s", pod.Name, pod.Namespace)
+}
+```
+should be changed to
+```go
+func updatePod(pod *covev1.Pod) {
+   ...
+   klog.InfoS("updated pod", "pod", klog.KObj(pod))
+}
+```
+And
+```go
+func updatePod(pod *covev1.Pod) {
+   ...
+   klog.Infof("updated pod, uid: %s", pod.Uid)
+}
+```
+should be changed to
+```go
+func updatePod(pod *covev1.Pod) {
+   ...
+   klog.InfoS("updated pod", "pod", klog.KObj(pod))
+}
+```
+
+`klog.KObj` requires passing a kubernetes object (struct implementing `metav1.Object` interface). In situation where
+object is not available, we can use `klog.KRef`. Still it is suggested to rewrite code to use object pointer instead of
+strings where possible.
+
+```go
+func updatePod(podName, podNamespace string) {
+   ...
+   klog.InfoS("updated pod", "pod", klog.KRef(podNamespace, podName))
+}
+```
+
+For non-namespaced object we can pass empty string to namespace argument
+
+```go
+func updateNode(nodeName string) {
+   ...
+   klog.InfoS("updated node", "node", klog.KRef("", nodeName))
+}
+```
+
+### Compare results
+
+With introduction of structured functions log arguments will be formatted automatically instead of depending on the
+caller. This means that we can remove burden of picking format by caller and ensure greater log consistency, but during
+migration it's important to ensure that we avoid degradation of log quality. We should ensure that during migration we
+preserve properties like:
+* meaning
```

Comment:
What does "meaning" mean in this context?

Reply:
We should make sure that meaning of event that log is describing is preserved